src/konserve/filestore.clj

```diff
@@ -360,20 +360,22 @@
 
   PBinaryAsyncKeyValueStore
   (-bget [this key locked-cb]
-    (let [fn     (str (uuid key))
-          f      (io/file (str folder "/data/" fn))
-          res-ch (chan)]
+    (let [fn (str (uuid key))
+          f         (io/file (str folder "/data/" fn))
+          res-ch    (chan)]
       ;; migrate old schema
-      (when (and stale-binaries? (.exists (io/file (str folder "/B_" fn))))
-        (read-binary-old f res-ch folder fn key #(-bassoc this key (:input-stream %)))
-        (.delete (io/file (str folder "/B_" fn)))
-        (when (:fsync config)
-          (sync-folder folder)))
-      (go (read-binary f res-ch folder fn key locked-cb))))
+      (if (and stale-binaries? (.exists (io/file (str folder "/B_" fn))))
+        (async/thread
+          (do
+            (read-binary-old (io/file (str folder "/B_" fn)) folder fn key #(write-binary folder fn key (:input-stream %) config))
+            (write-edn-key serializer write-handlers read-handlers (str folder "/meta/") fn {:key key :format :binary} config)
+            (delete-entry (str "/B_" fn) folder config)
+            (read-binary f res-ch folder fn key locked-cb)))
```

Comment:
read-binary returns already a go channel, so you have to <! here and not wrap in a go block below.